netplan/cli/commands/try_command.py

```diff
@@ -0,0 +1,137 @@
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+'''netplan try command line'''
+
+import fcntl
+import os
+import termios
+import time
+import select
+import signal
+import sys
+
+from netplan.configmanager import ConfigManager
+import netplan.cli.utils as utils
+from netplan.cli.commands.apply import NetplanApply
+
+# Keep a timeout long enough to allow the network to converge, 60 seconds may
+# be slightly short given some complex configs, i.e. if STP must reconverge.
+INPUT_TIMEOUT = 120
+
+
+class NetplanTry(utils.NetplanCommand):
+
+    def __init__(self):
+        super().__init__(command_id='try',
+                         description='Try to apply a new netplan config to running '
+                                     'system, with automatic rollback',
+                         leaf=True)
+        self.configuration_changed = False
+        self.config_manager = ConfigManager()
+
+    def run(self):  # pragma: nocover (requires user input)
+        self.parser.add_argument('--config-file',
+                                 help='Apply the config file in argument on top of current configuration.')
+
+        self.func = self.command_try
+
+        self.parse_args()
+        self.run_command()
+
+    def command_try(self):  # pragma: nocover (requires user input)
+        try:
+            fd = sys.stdin.fileno()
+            old_flags = fcntl.fcntl(fd, fcntl.F_GETFL)
+            old_term = termios.tcgetattr(fd)
+
+            raw_term = termios.tcgetattr(fd)
+            raw_term[3] = raw_term[3] & ~termios.ICANON
+            raw_term[3] = raw_term[3] & ~termios.ECHO
+            termios.tcsetattr(fd, termios.TCSANOW, raw_term)
+
+            fcntl.fcntl(fd, fcntl.F_SETFL, old_flags | os.O_NONBLOCK)
+
+            # we really don't want to be interrupted while doing backup/revert operations
+            signal.signal(signal.SIGINT, self._signal_handler)
+
+            self.backup()
+            self.setup()
+
+            NetplanApply.command_apply()
+
+            print("Do you want to keep these settings?\n\n")
+            print("Press ENTER before timeout to accept the new configuration\n\n")
+            timeout_now = INPUT_TIMEOUT
+            while (timeout_now > 0):
+                print("Network changes will revert in {:>2} seconds".format(timeout_now), end='\r')
+                i, o, err = select.select([sys.stdin], [], [], 1)
+                try:
+                    sys.stdin.read()
```

Comment:
I'm confused by this code. Shouldn't you inspect i before calling this? Shouldn't you check that the user has indeed pressed enter and not some other key?